_posts/2022-07-01-argocd-image-updater-for-functions.md

```diff
@@ -0,0 +1,248 @@
+```
+
+```bash
+kubectl get functions -n openfaas-fn
+
+NAME             IMAGE
+email-notify     welteki/email-notify:0.1.0
+marketing-list   welteki/marketing-list:0.1.0
+```
+
+### Install and configure the ArgoCD Image Updater
+We install the updater as Kubernetes workload in the Argo CD namespace.
+
+```bash
+kubectl apply -n argocd  \
+    -f https://raw.githubusercontent.com/argoproj-labs/argocd-image-updater/stable/manifests/install.yaml
+```
+
+The openfaas-functions application needs to be correctly annotated for the Argo CD image Updater to know which application and images it should inspect and update.
+
+The following annotation were added to the openfaas-functions application template:
+```yaml
+annotations:
+    argocd-image-updater.argoproj.io/image-list: marketingList=welteki/marketing-list:~0.1
+    argocd-image-updater.argoproj.io/marketingList.helm.image-spec: marketingList.image
+```
+
+### Watch how functions get updated
+
+We use [`envsubst`-style templates](https://docs.openfaas.com/reference/yaml/#yaml-environment-variable-substitution) in the YAML stack to make the image tag configurable.
+```yaml
+  marketing-list:
+    lang: node17
+    handler: ./marketing-list
+    image: ${REGISTRY:-docker.io}/${REPO:-welteki}/marketing-list:${TAG:-dev}
+```
+
+To build and push a new version of the images run
+```bash
+TAG=0.1.4 faas-cli publish -f stack.yml
+```
+
+The logs for the Argo Image Updater show that it is considering 1 annotated application for update. Once it queries the registry for updated images it will detect the new version of our function image and update the openfaas-functions application to use the new version of the marketing-list function.
+```
+time="2022-06-14T20:33:41Z" level=info msg="Starting image update cycle, considering 1 annotated application(s) for update"
+time="2022-06-14T20:33:43Z" level=info msg="Setting new image to welteki/marketing-list:0.1.4" alias=marketingList application=openfaas-functions image_name=welteki/marketing-list image_tag=0.1.3 registry=
+time="2022-06-14T20:33:43Z" level=info msg="Successfully updated image 'welteki/marketing-list:0.1.3' to 'welteki/marketing-list:0.1.4', but pending spec update (dry run=false)" alias=marketingList application=openfaas-functions image_name=welteki/marketing-list image_tag=0.1.3 registry=
+time="2022-06-14T20:33:43Z" level=info msg="Committing 1 parameter update(s) for application openfaas-functions" application=openfaas-functions
+time="2022-06-14T20:33:43Z" level=info msg="Successfully updated the live application spec" application=openfaas-functions
+```
+
+## Conclusion
+We used the Image Updater to automatically update our functions whenever we pushed a new version to the image registry. A similar configuration can be used with the openfaas-operator app. It could watch ghcr.io for updates to the OpenFaaS core components and automatically update them to the latest allowed version.
```

Comment:
The conclusion seems short.

Give an overview of what you did

Why someone might want this?

What other materials / posts they should also review if they happen to be here considering this.

Reply:
Can you link to the Argo Image Updater repo on GitHub?